Interactivity API: Server Directive Processing Refactor

luisherranz opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

<!-- In a few words, what is the PR actually doing? -->

This is a whole refactoring of the server directive processing logic in preparation for WP 6.5.

_I'll open another PR and a Trac ticket in WordPress Core._

There are also a few small tweaks in the APIs:

- **wp_interactivity_state( $namespace, $state )**

  I renamed `wp_initial_state` to `wp_interactivity_state` because there is now another `wp_interactivity_config` function to populate the serializable configuration in the server. Also, prefixing it with `_interactivity_` made sense to know what state we are referring to.

- **wp_interactivity_config( $namespace, $config )**

  New function to serialize non-reactive configuration. It's also namespaced.

  Internally, I renamed `<script type="application/json" id="wp-interactivity-initial-state">` to `<script type="application/json" id="wp-interactivity-data">` because now it contains both the state and config. It serializes the state in the `state` prop and the config in the `config` prop.

- **wp_interactivity_process_directives( $html )**

  New function to process the directives found in an HTML string. _This should now be suitable for classic themes as well._

## Why?

<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Because the server directive processing will likely be included in WordPress 6.5 and therefore we need to accommodate and prepare the code.

## How?

<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

By refactoring everything into a couple of classes: one that controls everything related to the Interactivity API in the server and another one that extends the WP 6.5 version of `WP_HTML_Tag_Processor` to add the ability to get and set content between balanced tags.

## Testing Instructions

- Create an interactive block with `supports.interactivity` set to `true` and with some directives.
- Test that the rendered HTML contains the correctly modified markup.

## Additional comments

The `data-wp-each` processor is still missing. I'll work on that on a separate PR.

---

Kudos to @ockham, @dmsnell and @c4rl0sbr4v0 for their previous contributions to this server directive processing ❤️

luisherranz committed:
Initial SDP version

luisherranz committed:
Rename `initialState` to `state`

luisherranz commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,639 @@
+	 * @since 6.5.0
+	 */
+	public function register_script_modules() {
+		gutenberg_register_module(
```
Rename with `wp_register_script_module` once [this PR](https://github.com/WordPress/wordpress-develop/pull/5869) lands in core and [this PR](https://github.com/WordPress/gutenberg/pull/57778) lands in Guntenberg.

luisherranz replied:
Done in https://github.com/WordPress/gutenberg/pull/58066/commits/bdd0fad0de043e112fa13cd25dbc2fd6a83a5c42.

luisherranz commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,639 @@
+			'@wordpress/interactivity',
+			gutenberg_url( '/build/interactivity/index.min.js' ),
+			array(),
+			defined( 'GUTENBERG_VERSION' ) ? GUTENBERG_VERSION : get_bloginfo( 'version' )
```
Remove this in the WordPress backport.

luisherranz commented on lib/compat/wordpress-6.5/interactivity-api/interactivity-api.php:
```diff
@@ -0,0 +1,127 @@
+/**
+ * Interactivity API: Functions and hooks
+ *
+ * @package Gutenberg
```
```suggestion
 * @package WordPress
```

luisherranz replied:
Done in https://github.com/WordPress/gutenberg/pull/58066/commits/bcefd095988267115f2fee15cb080ba0e1292dab.

luisherranz commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api-directives-processor.php:
```diff
@@ -0,0 +1,162 @@
+ * Class used to iterate over the tags of an HTML string and help process the
+ * directive attributes.
+ */
+class WP_Interactivity_API_Directives_Processor extends Gutenberg_HTML_Tag_Processor_6_5 {
```
Rename `Gutenberg_HTML_Tag_Processor_6_5` with `WP_HTML_Tag_Processor` in WordPress.

@dmsnell can you confirm that this is correct?

dmsnell replied:
when we merge it into Core it needs to be `WP_HTML_Tag_Processor` but while it remains in Gutenberg/as long as it remains in Gutenberg it should stay `Gutenberg_HTML_Tag_Processor_6_5` (or whatever version tracks Core `trunk`)

luisherranz replied:
Ok, thank you.

luisherranz commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api-directives-processor.php:
```diff
@@ -0,0 +1,162 @@
+		$this->release_bookmark( $start_name );
+		$this->release_bookmark( $end_name );
+
+		$this->lexical_updates[] = new Gutenberg_HTML_Text_Replacement_6_5( $start, $end - $start, esc_html( $new_content ) );
```
Rename `Gutenberg_HTML_Text_Replacement_6_5` with `WP_HTML_Text_Replacement` in WordPress.

@dmsnell can you confirm that this is correct?

dmsnell replied:
it's correct.

luisherranz replied:
Ok, thanks Dennis.

luisherranz committed:
Fix incorrect package

luisherranz committed:
Properly remove the SDP during the e2e tests

luisherranz commented:
PHP unit tests are failing but they are not related:

- https://github.com/WordPress/gutenberg/pull/58069

sethrubenstein commented:
Sort of a modules question but pertinent to `wp_interactivity_process_directives`...
We have an endpoint called `/iframe` that uses `template_redirect` to intercept the template loading process and provide a simplified version of the content. Let's assume it's a very basic template, like this:

```PHP
wp_head();
show_admin_bar( false );
while ( have_posts() ) {
	the_post();
	$post_content = get_the_content();
	$post_content = apply_filters( 'the_content', $post_content );
	echo wp_sprintf(
		'<div class="prc-platform__iframe__content">%1$s</div>',
		$post_content,
	);
}
wp_reset_postdata();
<?php
wp_footer();
exit();
```

I was expecting `apply_filters('the_content')` to handle modules like scripts, but that doesn't seem to be the case. Seeing this PR I'm thinking I'll need `wp_interactivity_process_directives` to automatically enqueue `viewModules`. Is that correct?

luisherranz committed:
Merge branch 'trunk' into interactivity-api-server-directive-processing-refactoring

luisherranz commented:
> I'm thinking I'll need wp_interactivity_process_directives to automatically enqueue viewModules. Is that correct?

Hey Seth.

Things should happen in cascade. You should not have to do anything special:

- You run the `the_content` filter.
- `do_blocks` is attached to the `the_content` filter.
- `do_blocks` render the blocks.
- Each block should enqueue its scripts and modules present in `viewScript` and `viewModule`.
- Each block triggers its own `render_block_data` and `render_block` filters.
- `wp_interactivity_process_directives` is attached to the `render_block_data` and `render_block` filters and should process the directives.
- Once `the_content` returns, all the scripts/modules should be enqueued and all the directives should be processed.

If you are not seeing this behavior, would you mind opening an issue with the steps to reproduce it and pinging me and @sirreal so we can take a look? Thanks!

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api-directives-processor.php:
```diff
@@ -0,0 +1,162 @@
+class WP_Interactivity_API_Directives_Processor extends Gutenberg_HTML_Tag_Processor_6_5 {
+	/**
+	 * Returns the content between two balanced tags.
+	 *
```
if this _must_ be public in order for the Interactivity API to call it internally then we can leave this public, but we can add ` * @access private` to note that this is an internal method only

luisherranz replied:
Done in https://github.com/WordPress/gutenberg/pull/58066/commits/5d68f196f68b8aeaaebfc592f5c5c2fe1f25f540.

cbravobernal commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api-directives-processor.php:
```diff
@@ -0,0 +1,162 @@
+
+/**
+ * Class used to iterate over the tags of an HTML string and help process the
+ * directive attributes.
```
```suggestion
 * directive attributes.
 @access private
```

luisherranz replied:
Done in https://github.com/WordPress/gutenberg/pull/58066/commits/5d68f196f68b8aeaaebfc592f5c5c2fe1f25f540.

cbravobernal commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api-directives-processor.php:
```diff
@@ -0,0 +1,162 @@
+	 * @return array|null A pair of bookmarks, or null if there's no matching closing tag.
+	 */
+	private function get_balanced_tag_bookmarks() {
+		$i = 0;
+		while ( array_key_exists( 'start' . $i, $this->bookmarks ) ) {
+			++$i;
+		}
+		$start_name = 'start' . $i;
+
+		$this->set_bookmark( $start_name );
```
```suggestion
		static $i = 0;
		$start_name = 'start' . ++$i;
```

dmsnell replied:
note that we may still have what should be an extremely unlikely reason to check if the array key exists, because hypothetically something could set one of these bookmarks from the outside. we can use `isset()` safely because a bookmark won't exist in the array if it's not there.

probably find simply to omit the loop, but even if we keep it in the static var will help us avoid walking through all of the bookmarks all of the time since we won't have to start at zero every time.

luisherranz replied:
Done in https://github.com/WordPress/gutenberg/pull/58066/commits/4ca55a5504f0c0080ddf88ee745439bd3a8398c6. I also removed the bookmarks to make them less unlikely to collide.

cbravobernal commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api-directives-processor.php:
```diff
@@ -0,0 +1,162 @@
+			return null;
+		}
+
+		$i = 0;
+		while ( array_key_exists( 'end' . $i, $this->bookmarks ) ) {
+			++$i;
+		}
+		$end_name = 'end' . $i;
```
```suggestion
		$end_name = 'end' . ++$i;
```

luisherranz replied:
Done in https://github.com/WordPress/gutenberg/pull/58066/commits/4ca55a5504f0c0080ddf88ee745439bd3a8398c6.

cbravobernal commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api-directives-processor.php:
```diff
@@ -0,0 +1,162 @@
+	 *
+	 * @return bool Whether the current tag is void or not.
+	 */
+	public function is_void_element(): bool {
+		$tag_name = $this->get_tag();
+		$tag_name = strtoupper( null !== $tag_name ? $tag_name : '' );
+
+		return (
+			'AREA' === $tag_name ||
+			'BASE' === $tag_name ||
+			'BR' === $tag_name ||
+			'COL' === $tag_name ||
+			'EMBED' === $tag_name ||
+			'HR' === $tag_name ||
+			'IMG' === $tag_name ||
+			'INPUT' === $tag_name ||
+			'LINK' === $tag_name ||
+			'META' === $tag_name ||
+			'SOURCE' === $tag_name ||
+			'TRACK' === $tag_name ||
+			'WBR' === $tag_name
+		);
+	}
```
We can use `WP_HTML_Processor::is_void()`

luisherranz replied:
Done in https://github.com/WordPress/gutenberg/pull/58066/commits/614ef66d41acafbfb9b4c8d43374fb44bfddf3d0.

cbravobernal commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,639 @@
+			wp_print_inline_script_tag(
+				wp_json_encode(
+					array(
+						'config' => (object) $this->config_data,
+						'state'  => (object) $this->state_data,
```
Do we want the array keys if we don't have config_data but have state_data or viceversa?

cbravobernal replied:
If not, we could ship something like this (code by @dmsnell )

```
if  ( ! ( $has_state_data || $has_config_data ) ) {
     return;
}

$store = array();

if ( $has_state_data ) {
    $store['state'] = (object) $this->state_data;
}

if ( $has_config_data ) {
   $store['config'] = (object) $this->config_data;
}

wp_print_inline_script_tag(
    wp_json_encode( $store, … )
)
 …
 ```

luisherranz replied:
Done in https://github.com/WordPress/gutenberg/pull/58066/commits/3425e5d1ee0d9225bf2ce20d157d7f5fe4d586c9

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,639 @@
+		// It returns the original content if the HTML is unbalanced because
+		// unbalanced HTML is not safe to process. In that case, the Interactivity
+		// API runtime will update the HTML on the client side during the hydration.
+		return $unbalanced || 0 < count( $tag_stack ) ? $html : $p->get_updated_html();
```
It's not worth changing this PR, but I think that this will currently fail (because of unbalanced tags) if processing SVG or MathML elements which are self-closing. We will be handling this in the HTML API soon (probably in 6.6) and so if it's the case we'll be able to lean on that functionality when it comes.

luisherranz replied:
I agree we can merge this and investigate that later so I've added a task to [the Tracking Issue](https://github.com/WordPress/gutenberg/issues/56803) to review what happens with SVG and MathML elements before WP 6.5.

cbravobernal commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,639 @@
+	 * directive attribute value.
+	 *
+	 * If the value doesn't contain an explicit namespace, it returns the default
+	 * one. If the value contains a JSON instead of a reference path, the function
```
```suggestion
	 * one. If the value contains a JSON object instead of a reference path, the function
```

luisherranz replied:
Done in https://github.com/WordPress/gutenberg/pull/58066/commits/d4f584ae50ef07d61314a6f3395bcde6cf992ff1.

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,639 @@
+		}
+
+		$matches       = array();
+		$has_namespace = preg_match( '/^([\w\-_\/]+)::(.*)$/', isset( $directive_value ) ? $directive_value : '', $matches );
```
```php
// @todo verify that this doesn't explode on large values, such as an image data URI
```

dmsnell replied:
to avoid one potential catastrophic case we could end the regex pattern at `::`, since the following `(.*)$` cannot fail to match.

```php
// $directive_value _must_ be set here given the early-abort above, right?
$has_namespace = 1 === preg_match( '~^([\w/_-]+)::~', $directive_value, $namespace_match );
if ( $has_namespace ) {
	$default_namespace = $matches[1];
	$directive_value   = substr( $directive_value, strlen( $matches[0] ) );
}

$might_be_json_object = '{' === $directive_value[0];
if ( $might_be_json_object ) {
	$decoded = json_decode( $directive_value );

	// Should we use an affirmative test here like `is_array( $decoded )`
	if ( isset( $decoded ) && 'null' !== $decoded ) {
		$directive_value = $decoded;
	}
}

return array( $default_namespace, $directive_value );
```

luisherranz replied:
Improved in https://github.com/WordPress/gutenberg/pull/58066/commits/560a17ec69a7845fd8d66f4b1cf6b51a9786e831.

I didn't add this because we also rely on `json_decode` to turn boolean and number strings into booleans and numbers.

`$might_be_json_object = '{' === $directive_value[0];`

dmsnell replied:
is this documented somewhere? the reason I ask is because if the attribute value is `data-wp--something='"compartir"'` should the directive value be `compartir` or rejected?

it was my assumption from the examples that _either_ we have a string value describing a property in the store _or_ we have a JSON object.

luisherranz replied:
We _JSONify_ the values:

- `data-wp--something="compartir"` value is `"compartir"`.
- `data-wp--something="'compartir'"` value is `'compartir'`.
- `data-wp--something="true"` value is `true`.
- `data-wp--something="false"` value is `false`.
- `data-wp--something="null"` value is `null`.
- `data-wp--something="1"` value is `1` (not `"1"`).
- `data-wp--something='{ "a": 1 }'` value is `{ a: 1 }` or `array( 'a' => 1 )`

Does that make sense?

cbravobernal commented on lib/compat/wordpress-6.5/interactivity-api/interactivity-api.php:
```diff
@@ -0,0 +1,127 @@
+					return $content;
+				};
+
+				add_filter( 'render_block', $process_interactive_blocks, 10, 2 );
```
Is this line creating a queue of filters with the same callback, per root interactive block in the post_content?

Maybe we can add a `remove_filter` in line 38, inside the conditional.

luisherranz replied:
Good point. Added in https://github.com/WordPress/gutenberg/pull/58066/commits/56f37470c58984aac9f29a7aa5ba0fd8c2c641d3.

c4rl0sbr4v0 commented:
As @dmsnell commented while we were pair-reviewing this PR, multiline comments are not used in WordPress Core. I'll add a commit to replace them:
```
// Not
// permitted.
```

```
/*
 * This
 * is allowed.
 */
```

cbravobernal committed:
Update multiline comments to fit WP standards

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,672 @@
+		$attribute_value = $p->get_attribute( 'data-wp-interactive' );
+		$decoded_json    = is_string( $attribute_value ) && ! empty( $attribute_value )
+		? json_decode( $attribute_value, true )
+		: null;
```
nitpick, but these appear to be missing an indent level

luisherranz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/58066/commits/246317af7c1730d95fd079511bcef8609026e460.

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,672 @@
+		// Separates the namespace from the context JSON object.
+		list( $namespace_value, $decoded_json ) = is_string( $attribute_value ) && ! empty( $attribute_value )
+		? $this->extract_directive_value( $attribute_value, $namespace_value )
+		: array( $namespace_value, null );
```
nitpick, but these appear to be missing an indent level

luisherranz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/58066/commits/246317af7c1730d95fd079511bcef8609026e460.

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,672 @@
+			}
+		}
+
+		// If it doesn't find the style property, it adds it at the end of the list.
```
there's a distinction in behavior here that might be worth noting in the documentation: appending a style property to the end of the inline style may produce a different result than replacing one that's already there.

consider properties with multiple properties that can change the same thing, like border, padding, margin, etc…

```html
<span style="border-style: dashed; border: 3px solid red;"></span>

Replace existing style.
<span style="border-style: inset; border: 3px solid red;"></span>

Append to end.
<span style="border-style: dashed; border: 3px solid red; border-style: inset;"></span>
```

In this example, replacement has left the property unaffected while appending has set it to most closely match the intention of this function call, assuming this function call wants to force the border-style value.

Without a full inline CSS parser (not just the lexical tokens, but also the semantics of constructing the full property set) it gets tricky making tactical inline CSS edits like this.

luisherranz replied:
Good point. I've moved the modified properties to the end in https://github.com/WordPress/gutenberg/pull/58066/commits/8ef546f276b6583a6ec25a7c41dc02d9358beaa2, just in case.

dmsnell commented on packages/interactivity/src/directives.js:
```diff
@@ -177,9 +177,11 @@ export default () => {
-						// need deps because it only needs to do it the first time.
+						/*
+						 * This seems necessary because Preact doesn't change the class
+						 * names on the hydration, so we have to do it manually. It doesn't
```
WPCS: These and all other uses of "we" will need to be removed for inclusion in Core.

luisherranz replied:
Funny thing, this code is already in Core 😄

I'll review all the JS comments anyway for WP 6.5. Thanks, Dennis.

dmsnell replied:
🤦‍♂️

personally I like the "we" voice since it's actually us as a body of developers building this. WordPress is not sentient…

dmsnell commented on packages/interactivity/src/directives.js:
```diff
@@ -226,24 +230,29 @@ export default () => {
+						/*
+						 * Default value in browsers is `-1` and an empty string is
+						 * cast to `0` instead.
+						 */
```
at first I thought this applied to the remaining attributes, but I think it's limited to `tabindex`. if that's the case it could be nice having that in the comment.

```php
/*
 * The value for `tabindex` follows the parsing rules for an integer.
 * if that fails, or if the attribute isn't present, then the browsers
 * should "follow platform conventions to determine if the element
 * should be considered as a focusable area," practically meaning that
 * most elements get a default of `-1` (not focusable), but several
 * also get a default of `0` (focusable in order after all elements
 * with a positive `tabindex` value).
 *
 * @see https://html.spec.whatwg.org/#tabindex-value
 */
```

I'm not sure why this is relevant here, but the comment jumped out at me and led me to investigate further. is it needed at all? is there a confusion that a comment should clarify?

luisherranz replied:
That part was copy-pasted [from Preact codebase](https://github.com/preactjs/preact/blob/ea49f7a0f9d1ff2c98c0bdd66aa0cbc583055246/src/diff/props.js#L110-L129), so whoever copied it left the same comment 🙂

luisherranz replied:
I added your comment in https://github.com/WordPress/gutenberg/pull/58066/commits/d7e73b516207fedfc70e7b51630d540daf6247da.

luisherranz committed:
Add @access private to WP_Interactivity_API_Directives_Processor

luisherranz committed:
Merge remote-tracking branch 'origin/interactivity-api-server-directive-processing-refactoring' into interactivity-api-server-directive-processing-refactoring

luisherranz committed:
Remove array check and rename bookmarks in WP_Interactivity_API_Directives_Processor

luisherranz committed:
Dont print config/state if it's empty

luisherranz committed:
Add failing test for </ >

luisherranz committed:
Add missing `object` word

luisherranz committed:
Missing indentation on ternary conditionals

luisherranz committed:
Improve namespace extraction from directive value

luisherranz committed:
Move modified properties to the end in set_style_property

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -237,8 +244,8 @@ public function process_directives( string $html ): string {
-					? $directive_processor_prefixes_reversed
-					: $directive_processor_prefixes,
+				? $directive_processor_prefixes_reversed
+				: $directive_processor_prefixes,
```
is an autoformatter removing the leading indentation on these?

luisherranz replied:
No, I think this is me forgetting 😄

Fixed in https://github.com/WordPress/gutenberg/pull/58066/commits/c45893f4a5e5bf4dbd57f637baf98ac621307849.

luisherranz committed:
Replace tabindex comment

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,669 @@
+
+		// Replaces the value and namespace if there is a namespace in the value.
+		if ( 1 === preg_match( '/^([\w\-_\/]+)::./', $directive_value ) ) {
+			list($default_namespace, $directive_value) = explode( '::', $directive_value, 2 );
```
nice

luisherranz committed:
Remove filter after processing and bump priority

luisherranz committed:
Replace own is_void with HTML API one

luisherranz committed:
Missing ternary indentation

luisherranz committed:
Avoid early returns

Because they don't seem to work: https://github.com/WordPress/gutenberg/pull/57778#issuecomment-1907684335

luisherranz commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api-directives-processor.php:
```diff
@@ -134,30 +134,12 @@ private function next_balanced_closer(): bool {
-			'TRACK' === $tag_name ||
-			'WBR' === $tag_name
-		);
+		return Gutenberg_HTML_Processor_6_5::is_void( null !== $tag_name ? $tag_name : '' );
```
Replace `Gutenberg_HTML_Processor_6_5` with `WP_HTML_Processor` in WordPress Core.

luisherranz committed:
Merge branch 'trunk' into interactivity-api-server-directive-processing-refactoring

luisherranz committed:
Update wp_register_script_module function

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/interactivity-api.php:
```diff
@@ -23,23 +23,40 @@
 	function wp_interactivity_process_directives_of_interactive_blocks( $parsed_block ) {
 		static $root_interactive_block = null;
 
+		/**
```
sorry for the noise and the nitpick. I think this code will pass fine, but for future reference, these are PHPDoc comments and I think we only want multi-line comments, which start with `/*` and not with `/**`

luisherranz replied:
Ohhh, ok. I'm learning a lot about comments lately 🤷‍♂️😄

luisherranz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/58066/commits/c3c87d330b630ab75d1b4625d983d73b48c1848a.

luisherranz committed:
Replaces some incorrect docblock comments with block comments

luisherranz commented:
Thanks for the thorough review, @dmsnell and @c4rl0sbr4v0! 🎉

I've addressed all your feedback, and I think this should be ready to merge now.

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,668 @@
+			$decoded_json = json_decode( $directive_value, true );
+			if ( null !== $decoded_json || 'null' === $directive_value ) {
+				$directive_value = $decoded_json;
+			}
```
it'd still be nice to clarify what the expectations are here for sending non-object JSON. what happens if we store `"something"` as the value, or `false` or `true`, or `[{"o":4}, null, 3e6]`

luisherranz replied:
Sure! Done in https://github.com/WordPress/gutenberg/pull/58066/commits/6b62b970c7986e457e3bb3d848292a9072653dfe and added more edge cases in https://github.com/WordPress/gutenberg/pull/58066/commits/93eefeb37184cca608c23c641661f7031f0f9c45.

dmsnell replied:
thanks @luisherranz - I think I poorly worded what I was trying to convey. I'm putting myself into the shoes of someone who's wanting to understand the role that JSON data plays in the directives. maybe this is all documented somewhere else (in which case a link would be great).

in all of the iterations until now, the function's docblock has given examples of some JSON inputs and what it will do with those, but even as we add new examples, it still leaves open the question of the other kinds of JSON input not listed.

it would be great to somewhat comprehensively indicate what role the JSON plays. right now, it seems like we allow all JSON values except for `null` and `"null"`. I will therefore assume that I can store a JSON array like the value of `d` in `d = [null, null, null]` as the directive value. there's no indication in the documentation what this should produce. it's valid JSON but it's neither a JSON object, nor a number, nor the string `"null"` nor `"false"` nor `"true"`.

back to the comment, I see two ways of changing this that would clarify the expectations, both involve swapping out _some_ examples of JSON inputs and replacing them with the specification for it.

1. match the docblock to the existing code

```php
/*
 * The value may be one of the following:
 *
 *  - a namespace-prefixed path into the store, e.g. `otherPlugin::actions.foo`.
 *  - a dot-separated path without a namespace, e.g. `actions.foo`.
 *  - any JSON-serialized value except `null` and the string `"null"`, e.g. `{"isOpen": false}`.
 *
 * If provided an invalid value, or a value that cannot be decoded as JSON
 * the original value will pass through and fail to refer to any values from
 * the store.
 *
 * Non-string JSON values will be coerced to strings in the markup.
 */
```

2. match the code to the existing docblock

```php
$decoded_json = json_decode( $directive_value, /* Disambiguate arrays from objects. */ false );
if (
	"true" === $decoded_json ||
	"false" === $decoded_json ||
	is_object( $decoded_json ) ||
	is_numeric( $decoded_json )
) {
	$directive_value = is_object( $decoded_json )
		? (array) $decoded_json
		: (string) $decoded_json;
}
```

I _think_ that the second approach more closely matches what you were intending based on the function's explanation, examples, and tests. I may be totally confused in part because I don't understand what it means to "parse it and return the resulting array." What does it mean to pass `{"isOpen": false}`? or `otherPlugin::{"isOpen": false}`? would this be valid? `otherPlugin::[null]`?

please receive this comment in a tone of curiosity and good-will. I'm not trying to demand you change this, but it remains one of the more confusing points for me to understand and I'm sharing because I would imagine that others could potentially have similar questions. I also understand that this is an internal method and not the public interface, but I would still find it valuable to know what's supposed to happen if I am pulled in to debug an issue with this code and have a JSON input that isn't specified in the function's docs.

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,668 @@
+			$attribute_value = $p->get_attribute( 'data-wp-interactive' );
+			$decoded_json    = is_string( $attribute_value ) && ! empty( $attribute_value )
+			? json_decode( $attribute_value, true )
+			: null;
```
nitpick: indent

luisherranz replied:
Same here. This was already fixed in https://github.com/WordPress/gutenberg/commit/246317af7c1730d95fd079511bcef8609026e460.

luisherranz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/58066/commits/6cee3ba0d25a79b4b4e902c0054cd0b008da7176.

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,668 @@
+			 */
+			$namespace_stack[] = isset( $decoded_json['namespace'] )
+			? $decoded_json['namespace']
+			: end( $namespace_stack );
```
nitpick: indent

luisherranz replied:
This was already fixed in https://github.com/WordPress/gutenberg/commit/246317af7c1730d95fd079511bcef8609026e460, so maybe the autoformatter is removing that indent after all. I'll investigate when that was re-added.

luisherranz replied:
Oh, they all disappeared when I re-indented the code after removing the early `return`. My bad then. Fixed in https://github.com/WordPress/gutenberg/pull/58066/commits/6cee3ba0d25a79b4b4e902c0054cd0b008da7176.

dmsnell commented on lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -0,0 +1,668 @@
+			// Separates the namespace from the context JSON object.
+			list( $namespace_value, $decoded_json ) = is_string( $attribute_value ) && ! empty( $attribute_value )
+			? $this->extract_directive_value( $attribute_value, $namespace_value )
+			: array( $namespace_value, null );
```
nitpick: indent

luisherranz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/58066/commits/6cee3ba0d25a79b4b4e902c0054cd0b008da7176.

dmsnell reviewed (approved):
Thanks for all your work on this @luisherranz 

As a short note, the phrasing "It only processes one root…" reads a bit awkwardly to me throughout the comments. "Processes only one root…" or "This processes only one root…" in these contexts flows smoother. Just my two cents.

There could probably be more to address here, but nothing seems like it's blocking. Would be nice if you could clarify before merge the expected behavior for non-object JSON values as noted inline, but that's not critical.

luisherranz committed:
Fix indent ternaries (again)

cbravobernal reviewed (approved):
Amazing job done here. Huge props @luisherranz

luisherranz committed:
Add more info about how extract_directive_value behaves

luisherranz committed:
Add more edge cases

luisherranz committed:
Improve wp_interactivity_process_directives_of_interactive_blocks comment

luisherranz commented:
> As a short note, the phrasing "It only processes one root…" reads a bit awkwardly to me throughout the comments. "Processes only one root…" or "This processes only one root…" in these contexts flows smoother. Just my two cents.

That's better, thank you. Done in https://github.com/WordPress/gutenberg/pull/58066/commits/1954f45ce99cc3f77a0e60cd80ab61a6fe806d9b.

> Would be nice if you could clarify before merge the expected behavior for non-object JSON values as noted inline

Done in https://github.com/WordPress/gutenberg/commit/6b62b970c7986e457e3bb3d848292a9072653dfe.

---

Thanks again, folks! I'll wait until the tests pass and I'll merge this 🙂

getdave commented:
I noticed this PR was merged after I raised the [PHP Sync Tracking Issue for WP 6.5](https://github.com/WordPress/gutenberg/issues/57959) and has changed PHP files that _may_ need backporting to WP Core.

I saw it was mentioned in https://github.com/WordPress/gutenberg/issues/57959#issuecomment-1910002617 so it will need backporting.

luisherranz commented:
For reference, this is the backport PR and ticket:

- https://github.com/WordPress/wordpress-develop/pull/5953
- https://core.trac.wordpress.org/ticket/60356
